### PR TITLE
Update pricelist.py

### PR DESCRIPTION
**Pull request by realsaiko**
Described in Issue #4624
break should be in 'if price is not False:' condition, otherwise it checks only first rule and if the product not found in first rule it would return price false.

### CODE REVIEW

**Comment by gfcapalbo** [comment]
:+1:  Used in production, allows to correctly evaluate all rules.  had the same problem , without the modification odoo considers only the first rule and then breaks execution anyway.

**Comment by StefanRijnhart** [comment]
:+1: thanks for the fix. It is now proposed to OCB here: https://github.com/OCA/OCB/issues/224

**Comment by nle-odoo** [comment]
Hello realsaiko,


It seems good to me, but to use your commit, the CLA would need to be signed: https://github.com/odoo/odoo/blob/8.0/doc/cla/sign-cla.md

Thanks.

**Comment by realsaiko** [comment]
Ok, I have signed for our company

**Comment by nle-odoo** [comment]
Thanks.

But after some review, even if the current behaviour (introduced with commit 20bbabc82) is not very good and is different than in 7.0, we can't make this change in stable 8.0 since it could mean change in prices for existing sale order in a stable version.

Currently we have:

* 7.0: stop at the first matching record rule which if based on a pricelist must also match,
* 8.0: stop at the first matching record rule (without seeing if an eventual base pricelist match),
* master: stop after the first matching record rule, if it is based on a non-matching pricelist, the price is the original price.

So if we have:

* pricelist a : rule for "price / 2" on imac,
* pricelist b : rule for "price * 10" on software,
* pricelist x : rule based on "pricelist a" without restriction, rule based on "pricelist b" without restriction,

If we add an imac in a sale order with "pricelist x":

* in 7.0: price of "imac" is halved => first matching record rule and the based pricelist also match,
* in 8.0: price of "imac" is halved => first matching record rule and the based pricelist also match,
* in master: price of "imac" is halved => first matching record rule and the based pricelist also match.
          
If we add "windows 8 cd":
          
* in 7.0: price of "windows 8 cd" is multiplied by 10 => we get to the second rule since the first one based pricelist didn't match,
* in 8.0: price of "windows 8 cd" is 0 => we get to the first matching but the based pricelist didn't match,
* in master: price of "windows 8 cd" is the original one => first rule match, but the based one uppon it doesn't match (so we get the original price),
             
If we add "ice cream" (not corresponding to any rule):

* in 7.0: price of "ice cream" is 0 => don't correspond to any rule (there should be a wildcard rule for the pricelist if this behaviour was not wanted),
* in 8.0: price of "ice cream" is 0 => the first rule match but the pricelist it is based uppon don't match,
* in master: price of "ice cream" is the original price => first rule match, but the based one uppon it doesn't match.

If the behaviour is to be fixed, it should be in master. But since commit https://github.com/odoo/odoo/commit/2ebbcba3 and the change of behavior (now instead of 0 when a pricelist doesn't match we get the original price) the price is never false and we will break with or without the commit of this pull request.
 
So since I can't merge this in 8.0 for stability, and in master since it wouldn't have an effect, I am closing this pull request.

Still, thanks for your contribution.

**Comment by realsaiko** [comment]
Thank you for your description.
I`ve checked master and 8.0 and yes, the computation in master is different and it might work (not checked on working environment), but I would like to disagree on version 8.0. You say that currently on 8.0 we have:
'8.0: stop at the first matching record rule (without seeing if an eventual base pricelist match)'

but I say that currently we have:
stop at the first record rule

because 'for rule in items:' has the 'break' at the end and it doesn`t matter if the price is false or not. My suggestion was to push break after check 'if price is not False:' so then we can have:
stop at the first MATCHING record rule

Then we go through items in rule and if price is not False we have a break, so we don`t check other items.
Currently it takes only first item in rule and stops without checking other items in rule, so there is no point in 'for rule in items:' loop.

Hopefully you understand what I'm saying and I've just checked the current code for 8.0 before commenting.

**Comment by nle-odoo** [comment]
There is seven continue in this extract of code : [addons/product/pricelist.py#L274-L314](https://github.com/odoo/odoo/blob/78568b59ab03988d2c8156168cab900750d39d2f/addons/product/pricelist.py#L274-L314).

If the rule doesn't match we get to the next one (via the continue), till the first one that match.

In the given case, the rule match but the issue is that the pricelist uppon which the rule is based doesn't match, because the pricelist has other constraints the first rule didn't have, for example the product or product category in the related issue.

In 7.0 if the based uppon pricelist didn't match, the break was where your commit moves it so we would get to the next rule.

In 8.0 probably as a dubious optimisation, we break even if the pricelist the rule is based uppon doesn't match => price is zero.

I as well as [dle](https://github.com/beledouxdenis) (master of bug fixing) find the behaviour of 7.0 better and that it would be good to get it back in 8.0, but for a [stable](https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean) version we can't make a change like this that may change existing prices in purchase or sale order.

So dle said that if we accepted the commit, it should be in master, but the commit doesn't apply to master since there is a change of behaviour in master that would render it without effect (the matching is the same as in 8.0 but if no pricelist match, the price is the original price (and not False) so the price is (i didn't checked everything so it is more a big probability) never False).

In conclusion, if the behaviour has to be corrected it should unfortunately be done in master.
